File: java-service/src/test/java/com/example/unit/StringServiceTest.java
package com.example.unit;

import com.example.StringService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("StringService Unit Tests")
class StringServiceTest {

    private StringService stringService;

    @BeforeEach
    void setUp() {
        stringService = new StringService();
    }

    @Nested
    @DisplayName("Reverse Tests")
    class ReverseTests {

        @Test
        @DisplayName("Should reverse a string correctly")
        void shouldReverseStringCorrectly() {
            // Given
            String input = "hello";

            // When
            String result = stringService.reverse(input);

            // Then
            assertEquals("olleh", result);
        }

        @Test
        @DisplayName("Should handle empty string")
        void shouldHandleEmptyString() {
            // Given
            String input = "";

            // When
            String result = stringService.reverse(input);

            // Then
            assertEquals("", result);
        }

        @Test
        @DisplayName("Should handle single character")
        void shouldHandleSingleCharacter() {
            // Given
            String input = "a";

            // When
            String result = stringService.reverse(input);

            // Then
            assertEquals("a", result);
        }

        @Test
        @DisplayName("Should handle null input")
        void shouldHandleNullInput() {
            // Given
            String input = null;

            // When
            String result = stringService.reverse(input);

            // Then
            assertNull(result);
        }

        @ParameterizedTest
        @CsvSource({
            "hello, olleh",
            "world, dlrow",
            "12345, 54321",
            "racecar, racecar",
            "A, A"
        })
        @DisplayName("Should reverse various strings")
        void shouldReverseVariousStrings(String input, String expected) {
            // When
            String result = stringService.reverse(input);

            // Then
            assertEquals(expected, result);
        }
    }

    @Nested
    @DisplayName("Palindrome Tests")
    class PalindromeTests {

        @Test
        @DisplayName("Should identify palindrome correctly")
        void shouldIdentifyPalindromeCorrectly() {
            // Given
            String input = "racecar";

            // When
            boolean result = stringService.isPalindrome(input);

            // Then
            assertTrue(result);
        }

        @Test
        @DisplayName("Should identify non-palindrome correctly")
        void shouldIdentifyNonPalindromeCorrectly() {
            // Given
            String input = "hello";

            // When
            boolean result = stringService.isPalindrome(input);

            // Then
            assertFalse(result);
        }

        @Test
        @DisplayName("Should handle palindrome with spaces and punctuation")
        void shouldHandlePalindromeWithSpacesAndPunctuation() {
            // Given
            String input = "A man, a plan, a canal: Panama";

            // When
            boolean result = stringService.isPalindrome(input);

            // Then
            assertTrue(result);
        }

        @Test
        @DisplayName("Should handle null input")
        void shouldHandleNullInput() {
            // Given
            String input = null;

            // When
            boolean result = stringService.isPalindrome(input);

            // Then
            assertFalse(result);
        }

        @ParameterizedTest
        @ValueSource(strings = {"racecar", "level", "deed", "noon", "civic", "radar"})
        @DisplayName("Should identify various palindromes")
        void shouldIdentifyVariousPalindromes(String input) {
            // When
            boolean result = stringService.isPalindrome(input);

            // Then
            assertTrue(result);
        }

        @ParameterizedTest
        @ValueSource(strings = {"hello", "world", "test", "java", "programming"})
        @DisplayName("Should identify various non-palindromes")
        void shouldIdentifyVariousNonPalindromes(String input) {
            // When
            boolean result = stringService.isPalindrome(input);

            // Then
            assertFalse(result);
        }
    }

    @Nested
    @DisplayName("Count Vowels Tests")
    class CountVowelsTests {

        @Test
        @DisplayName("Should count vowels correctly")
        void shouldCountVowelsCorrectly() {
            // Given
            String input = "hello";

            // When
            int result = stringService.countVowels(input);

            // Then
            assertEquals(2, result);
        }

        @Test
        @DisplayName("Should handle string with no vowels")
        void shouldHandleStringWithNoVowels() {
            // Given
            String input = "rhythm";

            // When
            int result = stringService.countVowels(input);

            // Then
            assertEquals(0, result);
        }

        @Test
        @DisplayName("Should handle null input")
        void shouldHandleNullInput() {
            // Given
            String input = null;

            // When
            int result = stringService.countVowels(input);

            // Then
            assertEquals(0, result);
        }

        @ParameterizedTest
        @CsvSource({
            "hello, 2",
            "world, 1",
            "aeiou, 5",
            "rhythm, 0",
            "programming, 3",
            "TEST, 1"
        })
        @DisplayName("Should count vowels in various strings")
        void shouldCountVowelsInVariousStrings(String input, int expected) {
            // When
            int result = stringService.countVowels(input);

            // Then
            assertEquals(expected, result);
        }
    }

    @Nested
    @DisplayName("Title Case Tests")
    class TitleCaseTests {

        @Test
        @DisplayName("Should convert to title case correctly")
        void shouldConvertToTitleCaseCorrectly() {
            // Given
            String input = "hello world";

            // When
            String result = stringService.toTitleCase(input);

            // Then
            assertEquals("Hello World", result);
        }

        @Test
        @DisplayName("Should handle single word")
        void shouldHandleSingleWord() {
            // Given
            String input = "hello";

            // When
            String result = stringService.toTitleCase(input);

            // Then
            assertEquals("Hello", result);
        }

        @Test
        @DisplayName("Should handle empty string")
        void shouldHandleEmptyString() {
            // Given
            String input = "";

            // When
            String result = stringService.toTitleCase(input);

            // Then
            assertEquals("", result);
        }

        @Test
        @DisplayName("Should handle null input")
        void shouldHandleNullInput() {
            // Given
            String input = null;

            // When
            String result = stringService.toTitleCase(input);

            // Then
            assertNull(result);
        }

        @ParameterizedTest
        @CsvSource({
            "hello world, Hello World",
            "JAVA PROGRAMMING, Java Programming",
            "test case, Test Case",
            "single, Single",
            "multiple   spaces, Multiple   Spaces"
        })
        @DisplayName("Should convert various strings to title case")
        void shouldConvertVariousStringsToTitleCase(String input, String expected) {
            // When
            String result = stringService.toTitleCase(input);

            // Then
            assertEquals(expected, result);
        }
    }

    @Nested
    @DisplayName("Remove Duplicates Tests")
    class RemoveDuplicatesTests {

        @Test
        @DisplayName("Should remove duplicate characters correctly")
        void shouldRemoveDuplicateCharactersCorrectly() {
            // Given
            String input = "hello";

            // When
            String result = stringService.removeDuplicates(input);

            // Then
            assertEquals("helo", result);
        }

        @Test
        @DisplayName("Should handle string with no duplicates")
        void shouldHandleStringWithNoDuplicates() {
            // Given
            String input = "world";

            // When
            String result = stringService.removeDuplicates(input);

            // Then
            assertEquals("world", result);
        }

        @Test
        @DisplayName("Should handle null input")
        void shouldHandleNullInput() {
            // Given
            String input = null;

            // When
            String result = stringService.removeDuplicates(input);

            // Then
            assertNull(result);
        }

        @ParameterizedTest
        @CsvSource({
            "hello, helo",
            "world, world",
            "programming, progamin",
            "test, test",
            "aaaaa, a",
            "abcabc, abc"
        })
        @DisplayName("Should remove duplicates from various strings")
        void shouldRemoveDuplicatesFromVariousStrings(String input, String expected) {
            // When
            String result = stringService.removeDuplicates(input);

            // Then
            assertEquals(expected, result);
        }
    }

    @Nested
    @DisplayName("Find Longest Word Tests")
    class FindLongestWordTests {

        @Test
        @DisplayName("Should find longest word correctly")
        void shouldFindLongestWordCorrectly() {
            // Given
            String input = "hello world programming";

            // When
            String result = stringService.findLongestWord(input);

            // Then
            assertEquals("programming", result);
        }

        @Test
        @DisplayName("Should handle single word")
        void shouldHandleSingleWord() {
            // Given
            String input = "hello";

            // When
            String result = stringService.findLongestWord(input);

            // Then
            assertEquals("hello", result);
        }

        @Test
        @DisplayName("Should handle empty string")
        void shouldHandleEmptyString() {
            // Given
            String input = "";

            // When
            String result = stringService.findLongestWord(input);

            // Then
            assertEquals("", result);
        }

        @Test
        @DisplayName("Should handle null input")
        void shouldHandleNullInput() {
            // Given
            String input = null;

            // When
            String result = stringService.findLongestWord(input);

            // Then
            assertEquals("", result);
        }

        @ParameterizedTest
        @CsvSource({
            "hello world programming, programming",
            "test case example, example",
            "java python javascript, javascript",
            "a bb ccc, ccc",
            "short medium long, medium"
        })
        @DisplayName("Should find longest word in various strings")
        void shouldFindLongestWordInVariousStrings(String input, String expected) {
            // When
            String result = stringService.findLongestWord(input);

            // Then
            assertEquals(expected, result);
        }
    }

    @Nested
    @DisplayName("Count Words Tests")
    class CountWordsTests {

        @Test
        @DisplayName("Should count words correctly")
        void shouldCountWordsCorrectly() {
            // Given
            String input = "hello world programming";

            // When
            int result = stringService.countWords(input);

            // Then
            assertEquals(3, result);
        }

        @Test
        @DisplayName("Should handle single word")
        void shouldHandleSingleWord() {
            // Given
            String input = "hello";

            // When
            int result = stringService.countWords(input);

            // Then
            assertEquals(1, result);
        }

        @Test
        @DisplayName("Should handle empty string")
        void shouldHandleEmptyString() {
            // Given
            String input = "";

            // When
            int result = stringService.countWords(input);

            // Then
            assertEquals(0, result);
        }

        @Test
        @DisplayName("Should handle null input")
        void shouldHandleNullInput() {
            // Given
            String input = null;

            // When
            int result = stringService.countWords(input);

            // Then
            assertEquals(0, result);
        }
    }

    @Nested
    @DisplayName("Anagram Tests")
    class AnagramTests {

        @Test
        @DisplayName("Should identify anagrams correctly")
        void shouldIdentifyAnagramsCorrectly() {
            // Given
            String text1 = "listen";
            String text2 = "silent";

            // When
            boolean result = stringService.isAnagram(text1, text2);

            // Then
            assertTrue(result);
        }

        @Test
        @DisplayName("Should identify non-anagrams correctly")
        void shouldIdentifyNonAnagramsCorrectly() {
            // Given
            String text1 = "hello";
            String text2 = "world";

            // When
            boolean result = stringService.isAnagram(text1, text2);

            // Then
            assertFalse(result);
        }

        @Test
        @DisplayName("Should handle anagrams with spaces")
        void shouldHandleAnagramsWithSpaces() {
            // Given
            String text1 = "conversation";
            String text2 = "voices rant on";

            // When
            boolean result = stringService.isAnagram(text1, text2);

            // Then
            assertTrue(result);
        }

        @Test
        @DisplayName("Should handle null inputs")
        void shouldHandleNullInputs() {
            // Given
            String text1 = null;
            String text2 = "test";

            // When
            boolean result = stringService.isAnagram(text1, text2);

            // Then
            assertFalse(result);
        }

        @ParameterizedTest
        @CsvSource({
            "listen, silent, true",
            "hello, world, false",
            "debit card, bad credit, true",
            "funeral, real fun, true",
            "school master, the classroom, true"
        })
        @DisplayName("Should test various anagram combinations")
        void shouldTestVariousAnagramCombinations(String text1, String text2, boolean expected) {
            // When
            boolean result = stringService.isAnagram(text1, text2);

            // Then
            assertEquals(expected, result);
        }
    }
}

File: node-service/tests/unit/stringService.test.js
/**
 * Jest unit tests for StringService
 */

const StringService = require('../../src/stringService');

describe('StringService', () => {
  let stringService;

  beforeEach(() => {
    stringService = new StringService();
  });

  describe('Reverse', () => {
    test('should reverse a string correctly', () => {
      const result = stringService.reverse('hello');
      expect(result).toBe('olleh');
    });

    test('should reverse a string correctly - INTENTIONALLY FAILING', () => {
      const result = stringService.reverse('hello');
      expect(result).toBe('hello'); // This will fail - should be 'olleh'
    });

    test('should handle empty string', () => {
      const result = stringService.reverse('');
      expect(result).toBe('');
    });

    test('should handle single character', () => {
      const result = stringService.reverse('a');
      expect(result).toBe('a');
    });

    test('should handle null input', () => {
      const result = stringService.reverse(null);
      expect(result).toBeNull();
    });

    test('should handle undefined input', () => {
      const result = stringService.reverse(undefined);
      expect(result).toBeNull();
    });

    test.each([
      ['hello', 'olleh'],
      ['world', 'dlrow'],
      ['12345', '54321'],
      ['racecar', 'racecar'],
      ['A', 'A']
    ])('should reverse "%s" to "%s"', (input, expected) => {
      const result = stringService.reverse(input);
      expect(result).toBe(expected);
    });
  });

  describe('Palindrome', () => {
    test('should identify palindromes correctly', () => {
      expect(stringService.isPalindrome('racecar')).toBe(true);
      expect(stringService.isPalindrome('level')).toBe(true);
    });

    test('should identify non-palindromes correctly', () => {
      expect(stringService.isPalindrome('hello')).toBe(false);
      expect(stringService.isPalindrome('world')).toBe(false);
    });

    test('should handle palindrome with spaces and punctuation', () => {
      expect(stringService.isPalindrome('A man, a plan, a canal: Panama')).toBe(true);
    });

    test('should handle null input', () => {
      expect(stringService.isPalindrome(null)).toBe(false);
    });

    test('should handle undefined input', () => {
      expect(stringService.isPalindrome(undefined)).toBe(false);
    });

    test.each([
      ['racecar', true],
      ['level', true],
      ['deed', true],
      ['noon', true],
      ['civic', true],
      ['radar', true],
      ['hello', false],
      ['world', false],
      ['test', false],
      ['java', false]
    ])('should identify "%s" as palindrome: %s', (input, expected) => {
      const result = stringService.isPalindrome(input);
      expect(result).toBe(expected);
    });
  });

  describe('Count Vowels', () => {
    test('should count vowels correctly', () => {
      const result = stringService.countVowels('hello');
      expect(result).toBe(2);
    });

    test('should handle string with no vowels', () => {
      const result = stringService.countVowels('rhythm');
      expect(result).toBe(0);
    });

    test('should handle null input', () => {
      const result = stringService.countVowels(null);
      expect(result).toBe(0);
    });

    test('should handle undefined input', () => {
      const result = stringService.countVowels(undefined);
      expect(result).toBe(0);
    });

    test.each([
      [2, 'hello'],
      [1, 'world'],
      [5, 'aeiou'],
      [0, 'rhythm'],
      [3, 'programming'],
      [1, 'TEST']
    ])('should count %i vowels in "%s"', (expected, input) => {
      const result = stringService.countVowels(input);
      expect(result).toBe(expected);
    });
  });

  describe('Title Case', () => {
    test('should convert to title case correctly', () => {
      const result = stringService.toTitleCase('hello world');
      expect(result).toBe('Hello World');
    });

    test('should handle single word', () => {
      const result = stringService.toTitleCase('hello');
      expect(result).toBe('Hello');
    });

    test('should handle empty string', () => {
      const result = stringService.toTitleCase('');
      expect(result).toBe('');
    });

    test('should handle null input', () => {
      const result = stringService.toTitleCase(null);
      expect(result).toBeNull();
    });

    test('should handle undefined input', () => {
      const result = stringService.toTitleCase(undefined);
      expect(result).toBeUndefined();
    });

    test.each([
      ['hello world', 'Hello World'],
      ['JAVA PROGRAMMING', 'Java Programming'],
      ['test case', 'Test Case'],
      ['single', 'Single'],
      ['multiple   spaces', 'Multiple   Spaces']
    ])('should convert "%s" to "%s"', (input, expected) => {
      const result = stringService.toTitleCase(input);
      expect(result).toBe(expected);
    });
  });

  describe('Remove Duplicates', () => {
    test('should remove duplicate characters correctly', () => {
      const result = stringService.removeDuplicates('hello');
      expect(result).toBe('helo');
    });

    test('should handle string with no duplicates', () => {
      const result = stringService.removeDuplicates('world');
      expect(result).toBe('world');
    });

    test('should handle null input', () => {
      const result = stringService.removeDuplicates(null);
      expect(result).toBeNull();
    });

    test('should handle undefined input', () => {
      const result = stringService.removeDuplicates(undefined);
      expect(result).toBeNull();
    });

    test.each([
      ['hello', 'helo'],
      ['world', 'world'],
      ['programming', 'progamin'],
      ['test', 'tes'],
      ['aaaaa', 'a'],
      ['abcabc', 'abc']
    ])('should remove duplicates from "%s" to get "%s"', (input, expected) => {
      const result = stringService.removeDuplicates(input);
      expect(result).toBe(expected);
    });
  });

  describe('Find Longest Word', () => {
    test('should find longest word correctly', () => {
      const result = stringService.findLongestWord('hello world programming');
      expect(result).toBe('programming');
    });

    test('should handle single word', () => {
      const result = stringService.findLongestWord('hello');
      expect(result).toBe('hello');
    });

    test('should handle empty string', () => {
      const result = stringService.findLongestWord('');
      expect(result).toBe('');
    });

    test('should handle null input', () => {
      const result = stringService.findLongestWord(null);
      expect(result).toBe('');
    });

    test('should handle undefined input', () => {
      const result = stringService.findLongestWord(undefined);
      expect(result).toBe('');
    });

    test.each([
      ['hello world programming', 'programming'],
      ['test case example', 'example'],
      ['java python javascript', 'javascript'],
      ['a bb ccc', 'ccc'],
      ['short medium long', 'medium']
    ])('should find longest word in "%s" as "%s"', (input, expected) => {
      const result = stringService.findLongestWord(input);
      expect(result).toBe(expected);
    });
  });

  describe('Count Words', () => {
    test('should count words correctly', () => {
      const result = stringService.countWords('hello world programming');
      expect(result).toBe(3);
    });

    test('should handle single word', () => {
      const result = stringService.countWords('hello');
      expect(result).toBe(1);
    });

    test('should handle empty string', () => {
      const result = stringService.countWords('');
      expect(result).toBe(0);
    });

    test('should handle null input', () => {
      const result = stringService.countWords(null);
      expect(result).toBe(0);
    });

    test('should handle undefined input', () => {
      const result = stringService.countWords(undefined);
      expect(result).toBe(0);
    });
  });

  describe('Anagram', () => {
    test('should identify anagrams correctly', () => {
      expect(stringService.isAnagram('listen', 'silent')).toBe(true);
      expect(stringService.isAnagram('debit card', 'bad credit')).toBe(true);
    });

    test('should identify non-anagrams correctly', () => {
      expect(stringService.isAnagram('hello', 'world')).toBe(false);
    });

    test('should handle null inputs', () => {
      expect(stringService.isAnagram(null, 'test')).toBe(false);
      expect(stringService.isAnagram('test', null)).toBe(false);
    });

    test('should handle undefined inputs', () => {
      expect(stringService.isAnagram(undefined, 'test')).toBe(false);
      expect(stringService.isAnagram('test', undefined)).toBe(false);
    });

    test.each([
      ['listen', 'silent', true],
      ['hello', 'world', false],
      ['debit card', 'bad credit', true],
      ['funeral', 'real fun', true],
      ['school master', 'the classroom', true]
    ])('should identify "%s" and "%s" as anagrams: %s', (text1, text2, expected) => {
      const result = stringService.isAnagram(text1, text2);
      expect(result).toBe(expected);
    });
  });

  describe('Capitalize Words', () => {
    test('should capitalize words correctly', () => {
      const result = stringService.capitalizeWords('hello world');
      expect(result).toBe('Hello World');
    });

    test('should handle null input', () => {
      const result = stringService.capitalizeWords(null);
      expect(result).toBeNull();
    });

    test('should handle undefined input', () => {
      const result = stringService.capitalizeWords(undefined);
      expect(result).toBeNull();
    });

    test.each([
      ['hello world', 'Hello World'],
      ['python programming', 'Python Programming'],
      ['test case', 'Test Case'],
      ['single', 'Single']
    ])('should capitalize "%s" to "%s"', (input, expected) => {
      const result = stringService.capitalizeWords(input);
      expect(result).toBe(expected);
    });
  });

  describe('Count Occurrences', () => {
    test('should count occurrences correctly', () => {
      const result = stringService.countOccurrences('hello world hello', 'hello');
      expect(result).toBe(2);
    });

    test('should handle null inputs', () => {
      expect(stringService.countOccurrences(null, 'test')).toBe(0);
      expect(stringService.countOccurrences('test', null)).toBe(0);
    });

    test('should handle undefined inputs', () => {
      expect(stringService.countOccurrences(undefined, 'test')).toBe(0);
      expect(stringService.countOccurrences('test', undefined)).toBe(0);
    });

    test.each([
      [2, 'hello', 'hello world hello'],
      [3, 'test', 'test test test'],
      [1, 'python', 'python programming'],
      [0, 'missing', 'no match']
    ])('should count %i occurrences of "%s" in "%s"', (expected, substring, text) => {
      const result = stringService.countOccurrences(text, substring);
      expect(result).toBe(expected);
    });
  });

  describe('Valid Email', () => {
    test('should validate emails correctly', () => {
      expect(stringService.isValidEmail('test@example.com')).toBe(true);
      expect(stringService.isValidEmail('user.name@domain.co.uk')).toBe(true);
    });

    test('should reject invalid emails', () => {
      expect(stringService.isValidEmail('invalid-email')).toBe(false);
      expect(stringService.isValidEmail('@domain.com')).toBe(false);
    });

    test('should handle null input', () => {
      expect(stringService.isValidEmail(null)).toBe(false);
    });

    test('should handle undefined input', () => {
      expect(stringService.isValidEmail(undefined)).toBe(false);
    });

    test.each([
      ['test@example.com', true],
      ['user.name@domain.co.uk', true],
      ['invalid-email', false],
      ['@domain.com', false],
      ['test@', false],
      ['test.domain.com', false]
    ])('should validate "%s" as email: %s', (email, expected) => {
      const result = stringService.isValidEmail(email);
      expect(result).toBe(expected);
    });
  });

  describe('Extract Numbers', () => {
    test('should extract numbers correctly', () => {
      const result = stringService.extractNumbers('hello 123 world 456');
      expect(result).toEqual([123, 456]);
    });

    test('should handle null input', () => {
      const result = stringService.extractNumbers(null);
      expect(result).toEqual([]);
    });

    test('should handle undefined input', () => {
      const result = stringService.extractNumbers(undefined);
      expect(result).toEqual([]);
    });

    test.each([
      ['hello 123 world 456', [123, 456]],
      ['test 1 2 3', [1, 2, 3]],
      ['no numbers here', []],
      ['42 is the answer', [42]]
    ])('should extract numbers from "%s" as %p', (text, expected) => {
      const result = stringService.extractNumbers(text);
      expect(result).toEqual(expected);
    });
  });

  describe('Remove Special Characters', () => {
    test('should remove special characters correctly', () => {
      const result = stringService.removeSpecialChars('hello@world!');
      expect(result).toBe('hello world');
    });

    test('should handle null input', () => {
      const result = stringService.removeSpecialChars(null);
      expect(result).toBeNull();
    });

    test('should handle undefined input', () => {
      const result = stringService.removeSpecialChars(undefined);
      expect(result).toBeNull();
    });

    test.each([
      ['hello@world!', 'hello world'],
      ['test#case$', 'test case'],
      ['no-special-chars', 'no special chars'],
      ['clean text', 'clean text']
    ])('should remove special chars from "%s" to "%s"', (input, expected) => {
      const result = stringService.removeSpecialChars(input);
      expect(result).toBe(expected);
    });
  });
});
